### PR TITLE
porting shoc_diag_moment_lbycond function

### PR DESCRIPTION
**Pull request by xyuan**
This new PR redo the porting of function shoc_diag_moment_lbycond.

https://github.com/E3SM-Project/scream/pull/533

[BFB]

### CODE REVIEW

File: components/cam/src/physics/cam/shoc.F90 [hunk 1]
```diff
@@ -1129,21 +1129,35 @@ subroutine diag_second_moments_lbycond(&
...
   do i=1,shcol
 
-    uf = sqrt(ustar2(i) + 0.3_rtype * wstar(i) * wstar(i))
+    uf = bfb_sqrt(ustar2(i) + 0.3_rtype * wstar(i) * wstar(i))
     uf = max(ufmin,uf)
 
     ! Diagnose thermodynamics variances and covariances
-    thl_sec(i) = 0.4_rtype * a_const * (wthl_sfc(i)/uf)**2
-    qw_sec(i) = 0.4_rtype * a_const * (wqw_sfc(i)/uf)**2
+    thl_sec(i) = 0.4_rtype * a_const * (wthl_sfc(i)/uf) * (wthl_sfc(i)/uf)
```
Comment (ambrad) [inline components/cam/src/physics/cam/shoc.F90:1159]: use bfb_square

File: components/cam/src/physics/cam/shoc.F90 [hunk 2]
```diff
@@ -1153,7 +1167,9 @@ subroutine diag_second_moments_lbycond(&
     wqw_sec(i) = wqw_sfc(i)
     uw_sec(i) = uw_sfc(i)
     vw_sec(i) = vw_sfc(i)
-    wtke_sec(i) = max(sqrt(ustar2(i)),0.01_rtype)**3
+
+    wtke0 = max(bfb_sqrt(ustar2(i)),0.01_rtype)
+    wtke_sec(i) = wtke0*wtke0*wtke0
```
Comment (ambrad) [inline components/cam/src/physics/cam/shoc.F90:1172]: use bfb_cube

**Review by ambrad — CHANGES_REQUESTED** [review]
@xyuan @singhbalwinder @tcclevenger @jeff-cohere don't forget the pattern that was established in P3: When needing to make BFB-related changes to the original F90, make one minimal PR to do that (so visual inspection is sufficient to OK the failed tests) and then a second BFB PR to do the C++ conversion.

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond.cpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#include "shoc_diag_second_moments_lbycond_impl.hpp"
+#include "share/scream_types.hpp"
+
+namespace scream {
+namespace shoc {
+
+/*
+ *  * Explicit instantiation for doing ice melting on Reals using the
```
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond.cpp:8]: Let's get rid of the "doing ice melting on Reals" comment here.

File: components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  // according to that of Andre et al. 1978.
+
+  const auto ufmin = 0.01;
+
+  auto uf = sqrt(ustar2+sp(0.3)*wstar*wstar);
+  uf = max(ufmin,uf);
+
+  // Diagnose thermodynamics variances and covariances
+  thl_sec   = sp(0.4)*sp(1.8)*(wthl_sfc/uf)*(wthl_sfc/uf);
+  qw_sec    = sp(0.4)*sp(1.8)*(wqw_sfc/uf)*(wqw_sfc/uf);
```
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:35]: Can someone remind me whether we advocate the use of `square` for expressions like `(wqw_sfc/uf)*(wqw_sfc/uf)` above in C++ (and `cube` for those like on line 46 below)?

Comment (xyuan) [reply components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:35]: @jeff-cohere @ambrad There is no square and cube function that take Scalar arguments, that is the reason why it is unrolled in the code.

Comment (ambrad) [reply components/scream/src/physics/shoc/shoc_diag_second_moments_lbycond_impl.hpp:35]: See https://github.com/E3SM-Project/scream/blob/b2349c327fc5f564860d3649b97dfe61a1a42b8f/components/cam/src/physics/cam/bfb_math.inc

**Comment by xyuan** [comment]
I close this PR, prefer redoing it instead of rebasing.

**Pull request closed without merging**